src/app/_components/NewsletterForm.tsx

```diff
@@ -0,0 +1,59 @@
+'use client'
+
+import React from 'react'
+
+import { zodResolver } from '@hookform/resolvers/zod'
+import { FieldErrors, useForm } from 'react-hook-form'
+import { z, ZodSchema } from 'zod'
+
+import { Button } from '@/components/Button'
+import ControlledInput from '@/components/ControlledInput'
+import Form from '@/components/Form'
+
+export const NewsletterSchema: ZodSchema = z.object({
+  email: z
+    .string()
+    .min(1, { message: 'This field has to be filled.' })
+    .email('This is not a valid email.'),
+})
+
+export type FormType = z.infer<typeof NewsletterSchema>
+
+export function NewsletterForm() {
+  const methods = useForm<FormType>({
+    resolver: zodResolver(NewsletterSchema),
+  })
+
+  const { isSubmitting } = methods.formState
+
+  async function onSubmit(values: FormType): Promise<void> {
+    await new Promise((resolve) => setTimeout(resolve, 2000))
+    methods.resetField('email')
+  }
+
+  function getError(
+    errors: FieldErrors<FormType>,
+    name: Extract<keyof FormType, string>,
+  ): string | undefined {
+    return errors[name]?.message as string | undefined
+  }
```

Comment:
```suggestion
  async function onSubmit(values: FormType) {
    await new Promise((resolve) => setTimeout(resolve, 2000))
    methods.resetField('email')
  }

  function getError(
    errors: FieldErrors<FormType>,
    name: Extract<keyof FormType, string>,
  ) {
    return errors[name]?.message
  }
```

As I shared with Mirha, I usually don't use return types for functions as TypeScript is able to infer them for us. Sometimes they are not inferrable and we need to add them manually, but not in this case.

Do you agree? Would you prefer to have them?